File: it/src/test/scala/org/zalando/nakadi/client/scala/ClientIntegrationTest.scala
package org.zalando.nakadi.client.scala
import scala.concurrent.Await
import scala.concurrent.duration.DurationInt

import org.scalatest.BeforeAndAfterAll
import org.scalatest.Matchers
import org.scalatest.WordSpec
import org.zalando.nakadi.client.scala.model.CompatibilityMode
import org.slf4j.LoggerFactory

class ClientIntegrationTest extends WordSpec with Matchers with BeforeAndAfterAll {
  import org.zalando.nakadi.client.scala.test.factory.events.MySimpleEvent._
  private val log = LoggerFactory.getLogger(this.getClass)
  private val client = ClientFactory.buildScalaClient()

  override def afterAll {
    ClientFactory.deleteEventTypesThatStartWith("ClientIntegrationTest")
    client.stop()
  }

  "create and get an eventType" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"ClientIntegrationTest-create-and-get-an-eventType"
    }
    val eventType = eventGenerator.eventType
    //POST
    val creationResult = Await.result(client.createEventType(eventType), 10.seconds)
    creationResult shouldBe None

    //GET
    val eventClientResult = Await.result(client.getEventType(eventType.name), 10.seconds)
    log.info("Result: ",eventClientResult)
    eventClientResult.isRight shouldBe true
    val Right(eventTypeOpt) = eventClientResult
    eventTypeOpt.isDefined shouldBe true
    val Some(eventTypeResult) = eventTypeOpt

    //    Compare EventType
    eventType.category shouldBe eventTypeResult.category
    eventType.dataKeyFields shouldBe List()
    eventType.name shouldBe eventTypeResult.name
    eventType.owningApplication shouldBe eventTypeResult.owningApplication
    eventType.partitionStrategy shouldBe eventTypeResult.partitionStrategy
    eventType.schema shouldBe eventTypeResult.schema
    eventType.statistics shouldBe eventTypeResult.statistics
    eventType.enrichmentStrategies shouldBe eventTypeResult.enrichmentStrategies
    eventType.partitionKeyFields shouldBe eventTypeResult.partitionKeyFields
    eventType.compatibilityMode shouldBe eventTypeResult.compatibilityMode
  }
  "update an existing eventType" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"ClientIntegrationTest-update-an-existing-eventType"
    }
    val eventType = eventGenerator.eventType

    //POST
    val creationResult = Await.result(client.createEventType(eventType), 10.seconds)
    creationResult shouldBe None

    //Check created event is there
    val createdResult = Await.result(client.getEventType(eventType.name), 10.seconds)
    createdResult.isRight shouldBe true

    val Right(optionalResult) = createdResult
    optionalResult.isDefined shouldBe true

    val Some(createdEvent) = optionalResult

    //Check that it is not equal to original
    eventType.owningApplication shouldBe createdEvent.owningApplication

    //The changed one
    createdEvent.name shouldBe eventType.name
    createdEvent.category shouldBe eventType.category
    createdEvent.dataKeyFields shouldBe null
    createdEvent.name shouldBe eventType.name
    createdEvent.owningApplication shouldBe eventType.owningApplication
    createdEvent.partitionStrategy shouldBe eventType.partitionStrategy
    createdEvent.schema shouldBe eventType.schema
    createdEvent.statistics shouldBe eventType.statistics
    createdEvent.enrichmentStrategies shouldBe eventType.enrichmentStrategies
    createdEvent.partitionKeyFields shouldBe eventType.partitionKeyFields
    createdEvent.compatibilityMode shouldBe eventType.compatibilityMode

    //UPDATE
    val changedEventType = eventType.copy(owningApplication = "Nakadi-klients(integration-test-suite)2", enrichmentStrategies = List(), compatibilityMode = Some(CompatibilityMode.FIXED))
    eventType.owningApplication should not be changedEventType.owningApplication

    val updateResult = Await.result(client.updateEventType(eventType.name, changedEventType), 10.seconds)

    updateResult.isEmpty shouldBe true //Error should be empty
    val eventClientResult = Await.result(client.getEventType(eventType.name), 10.seconds)
    eventClientResult.isRight shouldBe true
    val Right(eventTypeOpt) = eventClientResult
    eventTypeOpt.isDefined shouldBe true
    val Some(eventTypeResult) = eventTypeOpt

    //Check that it is not equal to original
    eventType.owningApplication should not be eventTypeResult.owningApplication

    //The changed one
    changedEventType.name shouldBe eventTypeResult.name
    changedEventType.category shouldBe eventTypeResult.category
    changedEventType.dataKeyFields shouldBe List()
    changedEventType.name shouldBe eventTypeResult.name
    changedEventType.owningApplication shouldBe eventTypeResult.owningApplication
    changedEventType.partitionStrategy shouldBe eventTypeResult.partitionStrategy
    changedEventType.schema shouldBe eventTypeResult.schema
    changedEventType.statistics shouldBe eventTypeResult.statistics
    changedEventType.enrichmentStrategies shouldBe eventTypeResult.enrichmentStrategies
    changedEventType.partitionKeyFields shouldBe eventTypeResult.partitionKeyFields
    changedEventType.compatibilityMode shouldBe changedEventType.compatibilityMode

  }

  "POST & GET & DELETE /event-types/{name}" in {
    val eventGenerator = new DefaultMySimpleEventGenerator() {
      def eventTypeId = s"ClientIntegrationTest-POST-GET-DELETE"
    }
    val eventType = eventGenerator.eventType
    //POST
    val creationResult = Await.result(client.createEventType(eventType), 10.seconds)
    creationResult shouldBe None

    //GET
    {
      val eventClientResult = Await.result(client.getEventType(eventType.name), 10.seconds)
      eventClientResult.isRight shouldBe true
      eventClientResult.right.get.isDefined shouldBe true
    }
    //DELETE
    val deletedResult = Await.result(client.deleteEventType(eventType.name), 10.seconds)
    deletedResult.isEmpty shouldBe true

    //GET
    val eventClientResult = Await.result(client.getEventType(eventType.name), 10.seconds)
    eventClientResult.isRight shouldBe true

    //Result is empty(404)
    eventClientResult.right.get.isDefined shouldBe false

  }

  "GET /metrics" in {
    val result = Await.result(client.getMetrics(), 10.seconds)
    result.isRight shouldBe true
    val Right(metricsOpt) = result
    metricsOpt.isDefined shouldBe true
    val Some(metrics) = metricsOpt
    println(metrics)
    metrics.version shouldNot be(null)
    metrics.gauges.size should be > 0
  }

  "GET /event-types" in {
    val result = Await.result(client.getEventTypes(), 10.seconds)
    result.isRight shouldBe true
    val Right(eventTypesOpt) = result
    eventTypesOpt.isDefined shouldBe true
    val Some(eventTypes) = eventTypesOpt
    eventTypes.size should (equal(0) or (be > 0))
  }

  "GET /registry/partitions-strategies" in {
    val result = Await.result(client.getPartitioningStrategies(), 10.seconds)
    result.isRight shouldBe true
  }

  "GET /registry/enrichment-strategies" in {
    val result = Await.result(client.getEnrichmentStrategies(), 10.seconds)
    result.isRight shouldBe true
    val Right(strategies) = result
    strategies.isDefined shouldBe true
  }

}

File: it/src/test/java/org/zalando/client/java/ClientIntegrationTest.java
package org.zalando.client.java;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;

import org.junit.After;
import org.junit.Test;
import org.zalando.nakadi.client.java.Client;
import org.zalando.nakadi.client.java.enumerator.EventEnrichmentStrategy;
import org.zalando.nakadi.client.java.enumerator.EventTypeCategory;
import org.zalando.nakadi.client.java.enumerator.PartitionStrategy;
import org.zalando.nakadi.client.java.model.EventType;
import org.zalando.nakadi.client.java.model.EventTypeSchema;
import org.zalando.nakadi.client.java.model.EventTypeStatistics;
import org.zalando.nakadi.client.java.model.Metrics;
import org.zalando.nakadi.client.java.test.event.generator.EventGenerator;
import org.zalando.nakadi.client.java.test.event.simple.MySimpleEventGenerator;
import org.zalando.nakadi.client.scala.ClientFactory;

public class ClientIntegrationTest {
    private Client client = ClientFactory.getJavaClient();

    @After
    public void shutdown() throws InterruptedException, ExecutionException {
        client.stop();
    }

    @Test
    public void createEventTypes() throws InterruptedException, ExecutionException {
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("ClientIntegrationTest-Java-postGetDeleteEventTypes").build();
        EventType originalEventType = gen.getEventType();

        // POST
        client.createEventType(originalEventType).get();

        // GET
        Optional<EventType> eventTypeResult = client.getEventType(originalEventType.getName()).get();
        assertTrue("Created Event should be returned", eventTypeResult.isPresent());
        EventType eventType = eventTypeResult.get();

        assertEquals(eventType.getCategory(), originalEventType.getCategory());
        assertEquals(eventType.getDataKeyFields(), originalEventType.getDataKeyFields());
        assertEquals(eventType.getName(), originalEventType.getName());
        assertEquals(eventType.getOwningApplication(), originalEventType.getOwningApplication());
        assertEquals(eventType.getPartitionKeyFields(), originalEventType.getPartitionKeyFields());
        assertEquals(eventType.getPartitionStrategy(), originalEventType.getPartitionStrategy());
        assertEquals(eventType.getSchema().getSchema(), originalEventType.getSchema().getSchema());
        assertEquals(eventType.getSchema().getType(), originalEventType.getSchema().getType());
        assertEquals(eventType.getStatistics(), originalEventType.getStatistics());

    }

    @Test
    public void updateEventTypes() throws InterruptedException, ExecutionException {
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("ClientIntegrationTest-Java-postGetDeleteEventTypes").build();
        EventType originalEventType = gen.getEventType();

        // POST
        client.createEventType(originalEventType).get();

        // GET
        Optional<EventType> eventTypeResult = client.getEventType(originalEventType.getName()).get();
        assertTrue("Created Event should be returned", eventTypeResult.isPresent());
        EventType eventType = eventTypeResult.get();

        assertEquals(eventType.getCategory(), originalEventType.getCategory());
        assertEquals(eventType.getDataKeyFields(), originalEventType.getDataKeyFields());
        assertEquals(eventType.getName(), originalEventType.getName());
        assertEquals(eventType.getOwningApplication(), originalEventType.getOwningApplication());
        assertEquals(eventType.getPartitionKeyFields(), originalEventType.getPartitionKeyFields());
        assertEquals(eventType.getPartitionStrategy(), originalEventType.getPartitionStrategy());
        assertEquals(eventType.getSchema().getSchema(), originalEventType.getSchema().getSchema());
        assertEquals(eventType.getSchema().getType(), originalEventType.getSchema().getType());
        assertEquals(eventType.getStatistics(), originalEventType.getStatistics());

        String name = eventType.getName();
        String owningApplication = "owningApplication";
        EventTypeCategory category = eventType.getCategory();
        List<EventEnrichmentStrategy> enrichmentStrategies = eventType.getEnrichmentStrategies();
        PartitionStrategy partitionStrategy = eventType.getPartitionStrategy();
        EventTypeSchema schema = eventType.getSchema();
        List<String> dataKeyFields = eventType.getDataKeyFields();
        List<String> partitionKeyFields = eventType.getPartitionKeyFields();
        EventTypeStatistics statistics = eventType.getStatistics();
        EventType changedEventType = new EventType(name, owningApplication, category, enrichmentStrategies,
                partitionStrategy, schema, dataKeyFields, partitionKeyFields, statistics);

        // Update
        client.updateEventType(originalEventType.getName(), changedEventType).get();
        Thread.sleep(3000);
        // GET
        eventTypeResult = client.getEventType(originalEventType.getName()).get();
        assertTrue("Created Event should NOT be returned", eventTypeResult.isPresent());
        assertEquals(eventTypeResult.get().getCategory(), originalEventType.getCategory());
        assertEquals(eventTypeResult.get().getDataKeyFields(), originalEventType.getDataKeyFields());
        assertEquals(eventTypeResult.get().getName(), originalEventType.getName());
        assertEquals(eventTypeResult.get().getOwningApplication(), changedEventType.getOwningApplication());
        assertEquals(eventTypeResult.get().getPartitionKeyFields(), originalEventType.getPartitionKeyFields());
        assertEquals(eventTypeResult.get().getPartitionStrategy(), originalEventType.getPartitionStrategy());
        assertEquals(eventTypeResult.get().getSchema().getSchema(), originalEventType.getSchema().getSchema());
        assertEquals(eventTypeResult.get().getSchema().getType(), originalEventType.getSchema().getType());
        assertEquals(eventTypeResult.get().getStatistics(), originalEventType.getStatistics());

    }

    @Test
    public void deleteEventTypes() throws InterruptedException, ExecutionException {
        EventGenerator gen = new MySimpleEventGenerator()//
                .withEventTypeId("ClientIntegrationTest-Java-postGetDeleteEventTypes").build();
        EventType originalEventType = gen.getEventType();

        // POST
        client.createEventType(originalEventType).get();

        // GET
        Optional<EventType> eventTypeResult = client.getEventType(originalEventType.getName()).get();
        assertTrue("Created Event should be returned", eventTypeResult.isPresent());
        EventType eventType = eventTypeResult.get();

        assertEquals(eventType.getCategory(), originalEventType.getCategory());
        assertEquals(eventType.getDataKeyFields(), originalEventType.getDataKeyFields());
        assertEquals(eventType.getName(), originalEventType.getName());
        assertEquals(eventType.getOwningApplication(), originalEventType.getOwningApplication());
        assertEquals(eventType.getPartitionKeyFields(), originalEventType.getPartitionKeyFields());
        assertEquals(eventType.getPartitionStrategy(), originalEventType.getPartitionStrategy());
        assertEquals(eventType.getSchema().getSchema(), originalEventType.getSchema().getSchema());
        assertEquals(eventType.getSchema().getType(), originalEventType.getSchema().getType());
        assertEquals(eventType.getStatistics(), originalEventType.getStatistics());

        // DELETE
        Void result = client.deleteEventType(originalEventType.getName()).get();

        // GET
        eventTypeResult = client.getEventType(originalEventType.getName()).get();
        assertFalse("Created Event should NOT be returned", eventTypeResult.isPresent());
    }

    @Test
    public void getMetrics() throws InterruptedException, ExecutionException {
        Optional<Metrics> result = client.getMetrics().get();
        assertTrue("Metrics should be returned", result.isPresent());
        Metrics metrics = result.get();
        assertNotNull("Version should be available", metrics.getVersion());
        assertTrue("Gauges should not be empty", metrics.getGauges().size() > 0);
    }

    @Test
    public void getEventTypes() throws InterruptedException, ExecutionException {
        Optional<List<EventType>> result = client.getEventTypes().get();
        List<EventType> events = result.get();
        assertTrue(result.isPresent());
        assertTrue(events.size() >= 0);
    }

    @Test
    public void getEnrichmentStrategies() throws InterruptedException, ExecutionException {
        Optional<List<EventEnrichmentStrategy>> result = client.getEnrichmentStrategies().get();
        assertTrue(result.isPresent());
        List<EventEnrichmentStrategy> enrichtmentStrategies = result.get();
        assertTrue("EventEnrichmentStrategy", enrichtmentStrategies.size() == 1);
    }

    @Test
    public void getPartitioningStrategies() throws InterruptedException, ExecutionException {
        Optional<List<PartitionStrategy>> result = client.getPartitioningStrategies().get();
        assertTrue(result.isPresent());
        List<PartitionStrategy> partitioningStrategies = result.get();
        assertTrue("PartitionStrategy", partitioningStrategies.size() == 3);
    }

}
